src/main/scala/ai/privado/tagger/S3Tagger.scala

```diff
@@ -0,0 +1,130 @@
+    val readBucketNames      = ListBuffer[String]()
+    val writeBucketNames     = ListBuffer[String]()
+    val readWriteBucketNames = ListBuffer[String]()
+
+    if (cpg.metaData.language.head == Languages.JAVASRC) {
+
+      // find all S3 Read tagged locations and extract bucket names
+      readBucketNames ++= cpg.call
+        .where(_.tag.valueExact("Storages.AmazonS3.Read"))
+        .repeat(_.astParent)(_.emit(_.isCall).maxDepth(4))
+        .isCallTo("bucket")
+        .argument(1)
+        .isLiteral
+        .map(bucket => {
+          val bucketName = bucket.code.stripPrefix("\"").stripSuffix("\"")
+          if bucketName.matches(bucketPattern) then bucketName
+          else ""
+        })
+        .filter(_.nonEmpty)
+        .l
+
+      // find all S3 Write tagged locations and extract bucket names
+      writeBucketNames ++= cpg.call
+        .where(_.tag.valueExact("Storages.AmazonS3.Write"))
+        .repeat(_.astParent)(_.emit(_.isCall).maxDepth(4))
+        .isCallTo("bucket")
+        .argument(1)
+        .isLiteral
+        .map(bucket => {
+          val bucketName = bucket.code.stripPrefix("\"").stripSuffix("\"")
+          if bucketName.matches(bucketPattern) then bucketName
+          else ""
+        })
+        .filter(_.nonEmpty)
+        .l
+
+    } else if (cpg.metaData.language.head == Languages.PYTHONSRC) {
+      // handle case where bucket names are assigned to identifier args having some "bucket" as name, eg.
+      //    s3_client.put_object(Body = pdf_result, Bucket = s3_bucket, Key = s3_filename)
+      val bucketArgs = cpg.call
+        .where(_.tag.valueExact("Storages.AmazonS3.ReadAndWrite"))
+        .argument
+        .isIdentifier
+        .name(".*bucket.*")
+        .l
+      // handle case where Bucket() call's arg is assigned bucket name, eg.
+      //    Bucket(MY_BUCKET_NAME)
+        ++ cpg.call
+          .where(_.tag.valueExact("Storages.AmazonS3.ReadAndWrite"))
+          .name("Bucket")
+          .argument
+          .isIdentifier
+          .whereNot(_.code(".*tmp.*"))
+          .l
+
+      val bucketNames = bucketArgs
+        .map(arg =>
+          cpg
+            .call(Operators.assignment)
+            .where(_.astChildren.isIdentifier.name(arg.name))
+            .argument
+            .argumentIndex(2)
+            .isLiteral
+            .code
+            .l
+        )
+        .l
+        .flatten
+        .distinct
+
+      readWriteBucketNames ++= bucketNames
+        .map(bucket => {
+          val bucketName =
+            bucket.stripPrefix("\"").stripSuffix("\"").stripPrefix("\'").stripSuffix("\'").stripPrefix("s3a://")
+          if bucketName.matches(bucketPattern) then bucketName
+          else ""
+        })
+        .filter(_.nonEmpty)
+        .l
+    }
+
+    // remove existing DB details from rule and add READ DB DETAILS to separate S3 DB details
+    DatabaseDetailsCache.removeDatabaseDetails(readRuleId)
```

Comment:
Any reason to remove this, do we suspect this getting added in DatabaseDetailsCache?

Reply:
Yes, since tagging has already run before, the DatabaseDetails already gets populated in the cache. We need to remove it once and add new info. This pass is run after source tagging is already done.